File: solutions/003.longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int result = 0;
        int[] toDiscardNum = new int[128];
        for(int i =0 ,j=0;j<s.length();j++){
            i = Math.max(toDiscardNum[s.charAt(j)],i);
            result = Math.max(result,j-i+1);
            toDiscardNum[s.charAt(j)] = j+1;
        
        }
        return result;
    
    }
}

File: solutions/003.longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    if(s.length === 0) return 0;
    const map = new Map();
    let res = 0;
    let left = 0, right = 0;
    for(let i = 0; i < s.length; i++){
        if(map.has(s[i])){
            res = Math.max(res, right - left + 1);
            left = Math.max(map.get(s[i]) + 1, left);
        }
        right = i;
        map.set(s[i], i);
    }
    res = Math.max(res, right - left + 1);
    return res;
};